src/compiler/checker.ts

```diff
@@ -7997,7 +8003,7 @@ namespace ts {
                     getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>accessExpression.argumentExpression).name)) :
                     undefined;
             if (propName !== undefined) {
-                const prop = getPropertyOfType(objectType, propName);
+                const prop = getPropertyOfType(objectType, propName, /*allowPartialUnions*/ true);

```

Comment:
This is a too permissive. For example, it will allow the following to compile with no error:

```ts
interface Foo {
    bar: { baz: string } | { qux: number }
}

function f(x: Foo) {
    return x['bar']['baz'];  // Should be an error
}
```

It's pretty easy to fix through:

```ts
const prop = getPropertyOfType(objectType, propName, !accessExpression);
```

This will allow partial unions only when the indexed access doesn't originate in an expression.

Reply:
Ah good catch. Fixed! Thanks for the review.